### DISCUSSION TITLE
RustLsp Command not found

### DISCUSSION THREAD

**Discussion by semi710** [mrcjkb/rustaceanvim#354] [Q&A] [answered]
### Have you read the docs and searched existing issues?

- [X] I have read [the README's troubleshooting guide](https://github.com/mrcjkb/rustaceanvim/blob/master/README.md#troubleshooting).
- [X] I have searched [existing issues](https://github.com/mrcjkb/rustaceanvim/issues?q=is%3Aopen+is%3Aissue+label%3Abug).
- [X] I have searched [existing discussions](https://github.com/mrcjkb/rustaceanvim/discussions).
- [ ] I have searched the existing issues of other plugins related to this issue (if applicable).
- [X] I have searched the [existing rust-analyzer issues](https://github.com/rust-lang/rust-analyzer/issues?q=is%3Aissue+) (if applicable).
- [X] I can reproduce the issue with stable Neovim (not a distribution or nightly) and the [minimal config](https://github.com/mrcjkb/rustaceanvim/blob/master/troubleshooting/minimal.lua).

### Neovim version (nvim -v)

v0.9.5

### Operating system/version

Arch Linux

### Output of :checkhealth rustaceanvim

```console
==============================================================================
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- WARNING dap not installed. Needed for debugging features [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap)

Checking external dependencies ~
- OK rust-analyzer: found 
- OK Cargo: found cargo 1.77.2 (e52e36006 2024-03-26)
- OK rustc: found rustc 1.77.2 (25ef9e3d8 2024-04-09)

Checking config ~
- OK No errors found in config.

Checking for conflicting plugins ~
- OK No conflicting plugins detected.

Checking for tree-sitter parser ~
- OK tree-sitter parser for Rust detected.
```

### How to reproduce the issue

```console
used the minimal config also `RustLsp` command not available.
```

### Expected behaviour

`RustLsp` Command should be available

### Actual behaviour

![image](https://github.com/mrcjkb/rustaceanvim/assets/60490474/e8efe70c-f367-442a-ac5a-a644285e59f8)

### The minimal config used to reproduce this issue.

```Lua
-- Minimal nvim config with lazy
-- Assumes a directory in $NVIM_DATA_MINIMAL
-- Start with
--
-- export NVIM_DATA_MINIMAL=$(mktemp -d)
-- export NVIM_APP_NAME="nvim-ht-minimal"
-- nvim -u NORC -u minimal.lua
--
-- Then exit out of neovim and start again.

-- Ignore default config
local config_path = vim.fn.stdpath('config')
vim.opt.rtp:remove(config_path)

-- Ignore default plugins
local data_path = vim.fn.stdpath('data')
local pack_path = data_path .. '/site'
vim.opt.packpath:remove(pack_path)

-- bootstrap lazy.nvim
data_path = assert(os.getenv('NVIM_DATA_MINIMAL'), '$NVIM_DATA_MINIMAL environment variable not set!')
local lazypath = data_path .. '/lazy/lazy.nvim'
local uv = vim.uv
  ---@diagnostic disable-next-line: deprecated
  or vim.loop
if not uv.fs_stat(lazypath) then
  vim.fn.system {
    'git',
    'clone',
    '--filter=blob:none',
    'git@github.com:folke/lazy.nvim.git',
    '--branch=stable',
    lazypath,
  }
end
vim.opt.rtp:prepend(lazypath)

local lazy = require('lazy')

lazy.setup({
  {
    'mrcjkb/rustaceanvim',
    version = '^4',
    init = function()
      -- Configure rustaceanvim here
      vim.g.rustaceanvim = {}
    end,
    ft = { 'rust' },
  },
  -- Add any other plugins needed to reproduce the issue.
  -- see https://github.com/folke/lazy.nvim#-lazynvim for details.
}, { root = data_path, state = data_path .. '/lazy-state.json', lockfile = data_path .. '/lazy-lock.json' })
```

**Comment by semi710** [comment]
![image](https://github.com/mrcjkb/rustaceanvim/assets/60490474/ad0444dc-7e8b-4652-94fa-c7e868cb2526)
i get this error when doing `Rust Analyzer start`

**Reply by mrcjkb** [reply, marked as answer]
Hey :wave:

The command becomes available when the LSP client attaches.
The health check doesn't print a version, which tells me you probably have the rustup wrapper for rust-analyzer installed, but not rust-analyzer (you can test this by trying to run rust-analyzer from the command line).

To install rust-analyzer, you need to install it with rustup: `rustup add component rust-analyzer`.

**Reply by semi710** [reply]
Hey 🍻
Thanks for the solution, buddy.
I am new to rust and didn't give a thought that I may need to add analyzer as component!

**Comment by Avi-D-coder** [comment]
I have a similar issue to this.
RA crashes on a particular file and I cannot see the log because RustLsp stops existing.

```
Client 1 quit with exit code 0 and signal 6                                                                       
E492: Not an editor command: RustLsp logFile 
```

**Reply by mrcjkb** [reply]
Hey 👋 

I suggest you try

- [the rustaceanvim troubleshooting guide](https://github.com/mrcjkb/rustaceanvim?tab=readme-ov-file#stethoscope-troubleshooting)
- [the rust-analyzer troubleshooting guide](https://rust-analyzer.github.io/manual.html#troubleshooting)

You can configure where to write the log file with `vim.g.rustaceanvim.server.logfile`

**Reply by Avi-D-coder** [reply]
Thanks, I did find that and get the log out with `vim.g.rustaceanvim.server.logfile = /tmp/rustaceanvim.log`.
It'd be nice if `RustLsp logFile` kept working after a client crash.

In my case the issue is a discrepancy between vscode rustaceanvim behavior with proc macros.
A bug report will follow eventually.

For now If anyone else reads this and is running into a crash that goes away when using `procMacro.enabled = false`, try ignoring macros until the client stays up. `procMacro.ignored = {["rasn-derive"] = { "AsnType" }}` did it for me.
